components/eventing-controller/testing/subscriber.go

```diff
@@ -0,0 +1,103 @@
+package testing
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/avast/retry-go"
+	pkgerrors "github.com/pkg/errors"
+)
+
+type Subscriber struct {
+	addr   string
+	t      *testing.T
+	server *http.Server
+}
+
+func NewSubscriber(addr string, t *testing.T) *Subscriber {
+	return &Subscriber{
+		addr: addr,
+		t:    t,
+	}
+}
+
+func (s *Subscriber) Start() {
+	store := ""
+	mux := http.NewServeMux()
+	mux.HandleFunc("/subscribe", func(w http.ResponseWriter, r *http.Request) {
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			s.t.Fatalf("failed to read data: %v", err)
+		}
+		store = string(data)
+	})
+	mux.HandleFunc("/check", func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte(store))
+		if err != nil {
+			s.t.Fatalf("failed to write to the response: %v", err)
+		}
+	})
+
+	s.server = &http.Server{
+		Addr:    s.addr,
+		Handler: mux,
+	}
+
+	go func() {
+		log.Printf("subscriber server is starting at %v", s.server.Addr)
+		err := s.server.ListenAndServe()
+		if err != nil {
+			if err != http.ErrServerClosed {
+				s.t.Fatalf("failed to start server: %v", err)
+			}
+		}
+	}()
+}
+
+func (s *Subscriber) Shutdown() {
+	go func() {
```

Comment:
@sayanh I've nothing against doing that in a goroutine, but what's the reason behind that?

Reply:
It is done this way to avoid waiting time for the shutdown when it is synchronous.